src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,249 @@
+  /**
+   * Number of items to be rendered at the end.
+   * Not recommended to go lower than the visible items in viewport.
+   * @default 20
+   */
+  renderAdditional?: number;
+};
+
+/**
+ * `VirtualScroll` component to render only visible items + some additional out of a big list.
+ * Good use in expected big number of data either in some of the components (e.g. Table) or just simple list of DOM elements.
+ * It has 2 wrapper elements, so expect DOM tree change. One is used for setting full expected height in the scrollable container
+ * and other is for transformation (translateY) to show the correct part of the list.
+ * Currently it works only when items direct parent is a scrollable container (body scroll is not supported yet).
+ * It supports only static (same) height rows virtualization. Expect some issues, if list consists of different height elements.
+ * @example
+ * const itemRenderer = React.useCallback(() => (
+ *  <div key={index}>
+ *    This is my item #{index}
+ *  </div>
+ * ), [])
+ * <VirtualScroll
+ *  itemsLength={1000}
+ *  itemRenderer={itemRenderer}
+ * />
+ */
+export const VirtualScroll = ({
+  itemsLength: childrenLength,
+  itemRenderer,
+  renderAdditional = 20,
+}: VirtualScrollProps) => {
+  const [startNode, setStartNode] = React.useState(0);
+  const [visibleNodeCount, setVisibleNodeCount] = React.useState(0);
+  const scrollContainer = React.useRef<HTMLElement>();
+  const childrenParentRef = React.useRef<HTMLDivElement>(null);
+  const childHeight = React.useRef(0);
+  const [translateY, setTranslateY] = React.useState(0);
+  const onScrollRef = React.useRef<(e: Event) => void>();
+
+  // Find scrollable parent
+  // Needed only on init
+  React.useLayoutEffect(() => {
+    const scrollableParent = getScrollableParent(childrenParentRef.current);
+    scrollContainer.current = scrollableParent;
+  }, []);
+
+  const visibleChildren = React.useMemo(() => {
+    const arr = [];
+    const endIndex = Math.min(
+      childrenLength,
+      startNode + visibleNodeCount + renderAdditional,
+    );
+    for (let i = startNode; i < endIndex; i++) {
+      arr.push(itemRenderer(i));
+    }
+    return arr;
+  }, [
+    childrenLength,
+    itemRenderer,
+    renderAdditional,
+    startNode,
+    visibleNodeCount,
+  ]);
+
+  // Get child height on init
+  React.useLayoutEffect(() => {
+    if (!childrenParentRef.current) {
+      return;
+    }
+    let heightSum = 0;
+    for (let i = startNode; i < visibleChildren.length + startNode; i++) {
+      const htmlElement = childrenParentRef.current.children.item(
+        Math.max(0, i - startNode),
+      ) as HTMLElement;
+      const elementHeight = getElementHeight(htmlElement);
+      heightSum += Number(elementHeight.toFixed(2));
+    }
+    const elementHeight = heightSum / visibleChildren.length;
+    childHeight.current = Math.ceil(elementHeight);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
```

Comment:
There may be no children on first render. The easiest solution is to ask the consumer to provide child height via a prop.

Reply:
Maybe, but the point is that I have 2nd implementation coming with dynamic heights, where I calculate those, so kinda wanted to have a base for that ready. I will think something about childHeight if there is no children...

Reply:
Ok, so changed.. What i did, added `visibleChildren.length` as dep to trigger calculation of child height and also guard to early quit if there are no visible children.
It runs initially (as only `renderAdditional` amount of elements are rendered) and then on the full render. Scroll does not influence this to be recalculated.